### ISSUE TITLE
`no-unnecessary-type-assertion`: two false positives where the assertion changes the declared type and typescript-eslint stays quiet

### ISSUE THREAD

**Issue by wagenet** [oxc-project/tsgolint#1212]
_Identified and written by Claude (Claude Code); reviewed by me before filing._

Two assertions that `tsgolint` reports as unnecessary, where removing them makes `tsc`
fail. In both cases `@typescript-eslint/no-unnecessary-type-assertion` on the same source
reports nothing, so these are divergences from the reference implementation rather than
cases where both agree.

I know this rule has a long false-positive history (#669, #1044, #1094, #1122, #1156, and
the open #1141). I checked these two against #1141 and they are a different mechanism —
see "What this is not" below.

## Environment

- oxlint 1.83.0
- tsgolint built from `main` (a0e243e)
- `@typescript-eslint/eslint-plugin` 8.67.0, ESLint 10.8.1, for the comparison
- TypeScript 6.0.3
- macOS arm64

`tsconfig.json`:

```json
{
  "compilerOptions": {
    "strict": true,
    "target": "ESNext",
    "module": "preserve",
    "moduleResolution": "bundler",
    "noEmit": true
  },
  "include": ["a.ts", "b.ts"]
}
```

`.oxlintrc.json`:

```json
{ "plugins": ["typescript"], "rules": { "typescript/no-unnecessary-type-assertion": "error" } }
```

## Case 1 — assertion widens a call's declared signature to accept a parameter

`a.ts`:

```ts
// `ActionReturn<void, R>` resolves to a zero-parameter function.
type ActionReturn<P, R> = P extends void ? () => Promise<R> : (payload: P) => Promise<R>;

declare function memberAction<P = void, R = object>(options: { path: string }): ActionReturn<P, R>;

export class Model {
	// The assertion widens a `() => Promise<object>` into one that accepts a payload.
	start = memberAction({ path: 'start' }) as (payload?: { flag?: boolean }) => Promise<object>;
}

export function callIt(m: Model) {
	return m.start({ flag: true });
}
```

```console
$ tsc --noEmit                 # exits 0
$ oxlint --type-aware a.ts
a.ts:8:10: error typescript(no-unnecessary-type-assertion): This assertion is unnecessary since it does not change the type of the expression.
$ eslint a.ts                  # @typescript-eslint/no-unnecessary-type-assertion: nothing
```

Remove the assertion and `tsc` fails:

```
a.ts(12,17): error TS2554: Expected 0 arguments, but got 1.
```

## Case 2 — assertion narrows `object` to a type with optional properties

`b.ts`:

```ts
interface WidgetMetrics {
	fetch(): void;
}

declare function requestsFor(): object | null;

export function refresh() {
	// The assertion is what makes `.widgetMetrics` legal; `object` has no such property.
	const requests = requestsFor() as { widgetMetrics?: WidgetMetrics } | null;
	requests?.widgetMetrics?.fetch();
}
```

```console
$ tsc --noEmit                 # exits 0
$ oxlint --type-aware b.ts
b.ts:9:19: error typescript(no-unnecessary-type-assertion): This assertion is unnecessary since it does not change the type of the expression.
$ eslint b.ts                  # @typescript-eslint/no-unnecessary-type-assertion: nothing
```

Remove the assertion and `tsc` fails:

```
b.ts(10,12): error TS2339: Property 'widgetMetrics' does not exist on type 'object'.
```

Both target types are ones the source type is loosely compatible with — a zero-arg function
is accepted where an optional-parameter signature is expected, and `object` is accepted
where a type with only optional properties is expected. But the assertion still changes the
expression's declared type, which is what the following line depends on.

## What this is not

I initially took case 1 for #1141, since `P` and `R` appear in none of `memberAction`'s
parameters. It is not:

- Giving `memberAction` a parameter that mentions `P` (`{ path: string; sample?: P }`), so
  the type argument is no longer inferable from the assertion's contextual type, does not
  change the result — still reported.
- It is deterministic here: 5/5 runs report it, and it does not need `--type-check`
  (#1141 reports only under `--type-check`, and varies run to run).

## A third case, which looks like upstream rather than tsgolint

Reducing case 1 to a non-generic signature also reports — but `@typescript-eslint` reports
it too, so this one is presumably a typescript-eslint bug rather than a tsgolint
divergence. Including it only so the boundary is clear; I have not reported it there.

```ts
declare function makeAction(options: { path: string }): () => Promise<object>;

export class Model {
	start = makeAction({ path: 'start' }) as (payload?: { flag?: boolean }) => Promise<object>;
}
export function callIt(m: Model) { return m.start({ flag: true }); }
```

Removing that assertion is also a `TS2554`.

## Why it matters

The fix for this rule is applied by plain `--fix`, so `oxlint --type-aware --fix` removes
both assertions and breaks the build. Migrating a monorepo from ESLint to oxlint, this rule
produced 8 findings across the repo and 2 of them were these, both load-bearing — the
other 6 were genuine. A 25% rate on an autofixable rule means `--fix` cannot be run
unattended for it.
